### ISSUE TITLE
Wrong suggestion for `redundant_closure_call` when body contains `return`, `?` or `try!`.

### ISSUE THREAD

**Issue by kennytm** [rust-lang-nursery/rust-clippy#1684]
Test case:

```rust
fn main() {
	let e = (|| -> Result<i32, i32> {
		Err(3)?;
		Ok(1)
	})();
	assert_eq(e, Err(3));
}
```

Clippy will suggest

```rust
warning: Try not to call a closure in the expression where it is declared.
 --> src/main.rs:2:10
  |
2 |   	let e = (|| -> Result<i32, i32> {
  |  __________^ starting here...
3 | | 		Err(3)?;
4 | | 		Ok(1)
5 | | 	})();
  | |_____^ ...ending here
  |
  = note: #[warn(redundant_closure_call)] on by default
help: Try doing something like: 
  | 	let e = {
  | 		Err(3)?;
  | 		Ok(1)
  | 	};
  = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#redundant_closure_call
```

But the suggestion is wrong, because the `?` will propagate the error outside the block. Similar problem for early-`return`.

Clippy should avoid generating `redundant_closure_call` if `?` is present in the body. 

After `catch` expression is stablized (rust-lang/rust#31436, `#![feature(catch_expr)]`), the lint could be re-enabled when the body only contains `?`. The suggestion should be `let e = catch { … }` though.

Not sure about early-`return` (including `try!`), maybe just turn off the code suggestion, and recommend the coder to restructure the expression.

---

```sh
$ rustc -vV
rustc 1.18.0-nightly (28a742997 2017-04-13)
binary: rustc
commit-hash: 28a74299778cdad4ea999e4ee8f8c1ef793338bd
commit-date: 2017-04-13
host: x86_64-apple-darwin
release: 1.18.0-nightly
LLVM version: 3.9

$ cargo -vV
cargo 0.19.0-nightly (c416fb60b 2017-04-05)
release: 0.19.0
commit-hash: c416fb60b11ecfd2a1ba0fb8567c9a92590b5d28
commit-date: 2017-04-05

$ cargo clippy -V
0.0.123
```
